src/core/qgstiledownloadmanager.cpp

```diff
@@ -50,21 +50,36 @@ void QgsTileDownloadManagerWorker::queueUpdated()
 
   if ( mManager->mShuttingDown )
   {
+    // here we HAVE to build up a list of replies from the queue before do anything
+    // with them. Otherwise we can hit the situation where aborting the replies
+    // triggers immediately their removal from the queue, and we'll be modifying
+    // mQueue elsewhere while still trying to iterate over it here => crash
+    QList< QNetworkReply * > replies;
+    replies.reserve( mManager->mQueue.size() );
     for ( auto it = mManager->mQueue.begin(); it != mManager->mQueue.end(); ++it )
     {
-      it->networkReply->abort();
+      replies << it->networkReply;
+    }
+    // now abort all replies
+    for ( QNetworkReply *reply : replies )
+    {
+      reply->abort();
     }
 
     quitThread();
     return;
   }
 
-  if ( mIdleTimer.isActive() && !mManager->mQueue.isEmpty() )
+  if ( mIdleTimer.isActive() && !mManager->mQueue.empty() )
   {
     // if timer to kill thread is running: stop the timer, we have work to do
     mIdleTimer.stop();
   }
 
+  // There's a potential race here -- if a reply finishes while we're still in the middle of iterating over the queue,
+  // then the associated queue entry would get removed while we're iterating over the queue here.
+  // So instead defer the actual queue removal until we've finished iterating over the queue.
```

Comment:
Hmm I am trying to understand how a race can happen here... if a reply finishes while we're iterating over the queue, I would expect that its QgsTileDownloadManagerReplyWorkerObject::replyFinished() handler gets called only when QgsTileDownloadManagerWorker::queueUpdated() returns (and the worker thread gets back to the event loop).

Or is it that `QgsNetworkAccessManager::instance()->get()` breaks my assumptions and inside it the event loop is entered prematurely?

Reply:
> Or is it that QgsNetworkAccessManager::instance()->get() breaks my assumptions and inside it the event loop is entered prematurely?

This. There's some circumstances in QgsNetworkAccessManager where we trigger an event loop, such as when an authentication method requires user entry, or when an ssl error occurs and the user is asked what to do.

Reply:
I see, that makes sense.